### PR TITLE
TRITON-567 would like metrics for node's GC and other overhead

### PR DESCRIPTION
**Pull request by joshwilsdon**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -17,6 +17,7 @@
     "artedi": "1.3.0",
     "assert-plus": "1.0.0",
     "bunyan": "1.8.12",
+    "event-loop-stats": "1.0.0",
     "vasync": "2.2.0",
```
Comment (dyep49) [inline package.json]: missing gc-stats

File: lib/metrics-manager.js [hunk 2]
```diff
@@ -134,6 +158,403 @@ MetricsManager.prototype.collectMetrics = function updateMetrics(name) {
...
+            // adding 0 to the collector every time, even if this type never
+            // happens again.
+            delete self.gcCounts[gcTypeNum];
+            delete self.gcPauses[gcTypeNum];
+        }
+
+        // FUTURE: Open FDs: readdir /proc/self/fd and count .length?
+
+        // Now actually collect the remaining data into the artedi metrics.
+        onEachCollector(function _createArtediCollector(type, name) {
```
Comment (dyep49) [inline lib/metrics-manager.js]: Nit: Maybe something like _updateArtediCollector is a more appropriate function name?

File: lib/metrics-manager.js [hunk 3]
```diff
@@ -134,6 +158,403 @@ MetricsManager.prototype.collectMetrics = function updateMetrics(name) {
...
+            nodejs_eventloop_max_tick_count: {
+                source: 'eventloopMetrics',
+                key: 'loopCountMax',
+                help: 'Maximum number of ticks in a single setTimeout() loop.'
+            },
+            nodejs_eventloop_max_time_seconds: {
+                source: 'eventloopMetrics',
+                key: 'loopTimeMax',
+                help:
+                    'Maxium number of seconds spent on a single eventloop ' +
```
Comment (dyep49) [inline lib/metrics-manager.js]: Nit: %s/Maxium/Maximum

File: lib/metrics-manager.js [hunk 4]
```diff
@@ -134,6 +158,403 @@ MetricsManager.prototype.collectMetrics = function updateMetrics(name) {
...
+            nodejs_active_requests_count: {
+                source: 'special',
+                help:
+                    'Total number of active requests reported by ' +
+                    'process._getActiveRequests().'
+            },
+            nodejs_eventloop_max_lag_seconds: {
+                source: 'eventloopMetrics',
+                key: 'lagMax',
+                help: 'Maxium number of seconds setTimeout() was delayed.'
```
Comment (dyep49) [inline lib/metrics-manager.js]: Nit: %s/Maxium/Maximum

File: lib/metrics-manager.js [hunk 5]
```diff
@@ -134,6 +158,403 @@ MetricsManager.prototype.collectMetrics = function updateMetrics(name) {
...
+            nodejs_eventloop_time_seconds_total: {
+                source: 'eventloopMetrics',
+                key: 'loopTimeSum',
+                help:
+                    'Total amount of time spent in the eventloop since ' +
+                    'start of observation.'
+            },
+            nodejs_gc_execution_count_total: {
+                source: 'special',
+                help: 'number of times node GC has run.'
```
Comment (dyep49) [inline lib/metrics-manager.js]: Super nit: %s/number/Number

**Review by dyep49 — COMMENTED** [review]
This looks great! I made a few tiny, tiny nits. Can't think of any other node/v8/event loop metrics that come to mind. I'll keep playing around with it today to see if that sparks any ideas. Do you have a feel for how much overhead this adds to the collection time?

**Comment by joshwilsdon** [comment]
Implemented with ad349ed3ea451797bd5e9cb9cb9bc52d92d51eda

**Pull request closed without merging**